File: Harvest/Services/ChatService.swift
import Foundation
import Supabase
import Realtime

struct ChatService {
    private var client: SupabaseClient { SupabaseManager.shared.client }

    func getMessages(conversationId: String) async throws -> [Message] {
        let messages: [Message] = try await client
            .from("messages")
            .select()
            .eq("conversation_id", value: conversationId)
            .order("created_at", ascending: true)
            .execute()
            .value
        return messages
    }

    func sendMessage(conversationId: String, senderId: String, content: String) async throws -> Message? {
        let now = ISO8601DateFormatter().string(from: Date())

        let messages: [Message] = try await client
            .from("messages")
            .insert([
                "conversation_id": conversationId,
                "sender_id": senderId,
                "content": content,
                "message_type": "text",
                "created_at": now
            ])
            .select()
            .execute()
            .value

        // Update conversation's last message
        try await client
            .from("conversations")
            .update([
                "last_message_at": now,
                "last_message_preview": String(content.prefix(100))
            ])
            .eq("id", value: conversationId)
            .execute()

        return messages.first
    }

    func ensureConversation(matchId: String, user1Id: String, user2Id: String) async throws -> String? {
        // Check existing
        let existing: [Conversation] = try await client
            .from("conversations")
            .select("id, user1_id, user2_id")
            .eq("match_id", value: matchId)
            .execute()
            .value

        if let existingConversation = existing.first {
            let normalizedUser1Id = user1Id.lowercased()
            let normalizedUser2Id = user2Id.lowercased()
            let needsRepair =
                existingConversation.user1Id?.lowercased() != normalizedUser1Id ||
                existingConversation.user2Id?.lowercased() != normalizedUser2Id

            if needsRepair {
                try await client
                    .from("conversations")
                    .update([
                        "user1_id": AnyJSON.string(user1Id),
                        "user2_id": AnyJSON.string(user2Id)
                    ])
                    .eq("id", value: existingConversation.id)
                    .execute()
            }

            return existingConversation.id
        }

        // Create new
        let created: [Conversation] = try await client
            .from("conversations")
            .insert([
                "match_id": matchId,
                "user1_id": user1Id,
                "user2_id": user2Id
            ])
            .select("id")
            .execute()
            .value

        return created.first?.id
    }

    func subscribeToMessages(conversationId: String, onMessage: @escaping @Sendable (Message) -> Void) -> RealtimeChannelV2 {
        let channel = client.realtimeV2.channel("messages:\(conversationId)")

        let changes = channel.postgresChange(
            InsertAction.self,
            table: "messages",
            filter: .eq("conversation_id", value: conversationId)
        )

        Task {
            for await change in changes {
                do {
                    let message = try change.decodeRecord(as: Message.self, decoder: JSONDecoder())
                    onMessage(message)
                } catch {
                    print("Warning: Failed to decode message from realtime update: \(error)")
                }
            }
        }

        Task {
            do {
                try await channel.subscribeWithError()
            } catch {
                print("Error: Failed to subscribe to messages channel: \(error)")
            }
        }

        return channel
    }

    func unsubscribe(channel: RealtimeChannelV2) {
        Task {
            await channel.unsubscribe()
        }
    }

    // MARK: - Read Receipts

    func markAsRead(messageId: String) async throws {
        let now = ISO8601DateFormatter().string(from: Date())
        try await client
            .from("messages")
            .update([
                "is_read": AnyJSON.bool(true),
                "read_at": AnyJSON.string(now)
            ])
            .eq("id", value: messageId)
            .execute()
    }

    // MARK: - Typing Indicators

    func sendTypingIndicator(conversationId: String, userId: String) async {
        let channel = client.realtimeV2.channel("typing:\(conversationId)")

        do {
            try await channel.subscribeWithError()
            try await Task.sleep(for: .milliseconds(200))
        } catch {
            print("Warning: Failed to setup typing indicator channel: \(error)")
            return // Don't broadcast if subscription failed
        }

        await channel.broadcast(
            event: "typing",
            message: ["user_id": AnyJSON.string(userId)]
        )

        Task {
            do {
                try await Task.sleep(for: .seconds(1))
            } catch {
                // Task cancelled, proceed with cleanup
            }
            await channel.unsubscribe()
        }
    }

    func subscribeToTyping(conversationId: String, onTyping: @escaping @Sendable (String) -> Void) -> RealtimeChannelV2 {
        let channel = client.realtimeV2.channel("typing:\(conversationId)")

        let broadcasts = channel.broadcastStream(event: "typing")

        Task {
            for await message in broadcasts {
                if let userId = message["user_id"]?.stringValue {
                    onTyping(userId)
                }
            }
        }

        Task {
            do {
                try await channel.subscribeWithError()
            } catch {
                print("Error: Failed to subscribe to typing indicators: \(error)")
            }
        }

        return channel
    }
}

File: android/app/src/main/java/com/harvestglass/harvest/data/service/ChatService.kt
package com.harvestglass.harvest.data.service

import com.harvestglass.harvest.data.model.Message
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Order
import io.github.jan.supabase.postgrest.query.filter.FilterOperator
import io.github.jan.supabase.realtime.PostgresAction
import io.github.jan.supabase.realtime.broadcastFlow
import io.github.jan.supabase.realtime.channel
import io.github.jan.supabase.realtime.decodeRecord
import io.github.jan.supabase.realtime.postgresChangeFlow
import io.github.jan.supabase.realtime.realtime
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.mapNotNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import java.time.Instant

/** Mirrors Harvest/Services/ChatService.swift. */
class ChatService(private val client: SupabaseClient) {

    suspend fun getMessages(conversationId: String): List<Message> =
        client.postgrest.from("messages")
            .select {
                filter { eq("conversation_id", conversationId) }
                order("created_at", Order.ASCENDING)
            }
            .decodeList()

    /**
     * Inserts the message and updates the conversation's preview columns, so
     * the Seeds list reflects the new activity without a separate refresh.
     */
    suspend fun sendMessage(conversationId: String, senderId: String, content: String): Message? {
        val now = Instant.now().toString()

        val inserted = client.postgrest.from("messages")
            .insert(
                buildJsonObject {
                    put("conversation_id", conversationId)
                    put("sender_id", senderId)
                    put("content", content)
                    put("message_type", "text")
                    put("created_at", now)
                }
            ) { select() }
            .decodeList<Message>()

        client.postgrest.from("conversations").update(
            buildJsonObject {
                put("last_message_at", now)
                put("last_message_preview", content.take(PREVIEW_LENGTH))
            }
        ) { filter { eq("id", conversationId) } }

        return inserted.firstOrNull()
    }

    suspend fun markAsRead(messageId: String) {
        client.postgrest.from("messages").update(
            buildJsonObject {
                put("is_read", true)
                put("read_at", Instant.now().toString())
            }
        ) { filter { eq("id", messageId) } }
    }

    /** Live inserts for one conversation, server-filtered by conversation_id. */
    fun subscribeToMessages(conversationId: String): Flow<Message> = flow {
        val channel = client.realtime.channel("messages:$conversationId")
        val changes = channel.postgresChangeFlow<PostgresAction.Insert>(schema = "public") {
            table = "messages"
            filter("conversation_id", FilterOperator.EQ, conversationId)
        }
        channel.subscribe()
        emitAll(changes.mapNotNull { runCatching { it.decodeRecord<Message>() }.getOrNull() })
    }

    /**
     * Typing uses a broadcast channel rather than postgres_changes — nothing
     * is persisted, so there is no row to listen to. Emits the sender's id.
     */
    fun subscribeToTyping(conversationId: String): Flow<String> = flow {
        val channel = client.realtime.channel("typing:$conversationId")
        val broadcasts = channel.broadcastFlow<JsonObject>(event = TYPING_EVENT)
        channel.subscribe()
        emitAll(
            broadcasts.mapNotNull { payload ->
                runCatching { payload[TYPING_USER_KEY]?.jsonPrimitive?.content }.getOrNull()
            }
        )
    }

    suspend fun sendTypingIndicator(conversationId: String, userId: String) {
        val channel = client.realtime.channel("typing:$conversationId")
        runCatching {
            channel.subscribe()
            channel.broadcast(
                event = TYPING_EVENT,
                message = buildJsonObject { put(TYPING_USER_KEY, userId) }
            )
        }
        // A failed broadcast is not worth surfacing: the indicator is a nicety
        // and the Swift version only logs it.
    }

    companion object {
        private const val PREVIEW_LENGTH = 100
        private const val TYPING_EVENT = "typing"
        private const val TYPING_USER_KEY = "user_id"
    }
}
